Add py:type directive and role for documenting type aliases (#11989)
sphinx-doc/sphinx: doc/usage/domains/python.rst

.. rst:directive:: .. py:data:: name

   Describes global data in a module, including both variables and values used
   as "defined constants."  Class and object attributes are documented
   with :rst:dir:`py:attribute`.  Type aliases are documented with

>    :rst:dir:`py:type`.

Reviewer: ```suggestion
   as "defined constants".  Consider using :rst:dir:`py:type` for type
   aliases instead and :rst:dir:`py:attribute` for class variables and
   instance attributes.
```

   Describes an object data attribute.  The description should include
   information about the type of the data to be expected and whether it may be

>    changed directly.  Type aliases are documented with :rst:dir:`py:type`.

Reviewer: ```suggestion
   changed directly.  Type aliases should be documented with :rst:dir:`py:type`.
```

      Describe the location where the object is defined.  The default value is
      the module specified by :rst:dir:`py:currentmodule`.

.. rst:directive:: .. py:type:: name

   Describes a :ref:`type alias <python:type-aliases>`.  A description of
   the type alias, such as the docstring can be, placed in the body of
   the directive.

   .. versionadded:: 7.3

   .. rubric:: options

   .. rst:directive:option:: type: the type that the alias represents
      :type: text

> 

Reviewer: ```suggestion
.. rst:directive:: .. py:type:: name

   Describe a :ref:`type alias <python:type-aliases>`.  
   
   This directive supports an optional description body, e.g.::
   
   .. code-block:: rst
   
      .. py:type:: UInt64
      
      	 Represent a 64-bit positive integer.

   .. rubric:: options

   .. rst:directive:option:: canonical
      :type: text
      
      The canonical type represented by this alias, e.g.::
 
      .. code-block:: rst
 
	     .. py:type:: StrPattern
	        :canonical: str | re.Pattern[str]
    
         	Represent a regular expression or a compiled pattern.
         	      
   .. versionadded:: 7.3
```

I suggested `canonical` because `type` is a bit weird in the sense it's not the type *of* the type but rather the underlying type. In addition, I am wondering whether the aliased type should be the fully-qualified name (in which case we wouldn't have any issues with references I think and we could have "clickable" items) or if it's simply a text (I'll probably comment below).

Reviewer: Would the type aliases be considered as clickable items in base classes / type annotations, etc and so? (i.e., do they behave similarly to a class?)

Author: Yes, they would be.
